sentry_streams/sentry_streams/dummy/dummy_adapter.py

```diff
@@ -10,26 +25,74 @@
 class DummyAdapter(StreamAdapter[DummyInput, DummyOutput]):
     """
     An infinitely scalable adapter that throws away all the data it gets.
+    The adapter tracks the 'streams' that each step of iterate_edges() returns in the form of
+    lists of previous step names.
     """
 
     def __init__(self, _: PipelineConfig) -> None:
-        pass
+        self.input_streams: MutableMapping[str, List[str]] = {}
+
+    def track_input_streams(self, step: WithInput) -> None:
+        """
+        Tracks the streams that each step receives as input.
+        This can be used in tests to verify that steps downstream from a split in
+        the stream (such as a Router) are being applied to the correct stream.
+
+        For example, if we have:
+        Source --> Router --> Branch1 --> Map1
+                        |
+                        --> Branch2 --> Map2
+
+        We can verify that:
+        input_streams["Map1"] == ["Source", "Router", "Branch1"]
+        input_streams["Map2"] == ["Source", "Router", "Branch2"]
+        """
+        # TODO: update to support multiple inputs to a step
+        # once we implement Union
+        assert (
+            len(step.inputs) == 1
+        ), "Only steps with a single input are supported for DummyAdapter."
+
+        input_step = step.inputs[0]
+        input_step_name = input_step.name
+        input_step_stream = self.input_streams[input_step_name]
+        self.input_streams[step.name] = input_step_stream + [input_step_name]
+        # if step is a Router, also add its Branch nodes to input_streams
+        if isinstance(step, Router):
+            for branch in step.routing_table.values():
+                self.input_streams[branch.name] = self.input_streams[step.name] + [step.name]
```

Comment:
Please avoid this isinstance. You do not need that.

Jsut put the Router specific logic into the `router` method after calling `track_input_stream` where it belongs.
If you already have multiple methods, each of which contain the responsibility to manage a step, but then always call the same general method that split the responsibilities again, the abstraction is working against you rather than for you.